File: backend/app/models/oauth.py
"""OAuth 関連モデル。"""

from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field, HttpUrl


class OAuthInitiateRequest(BaseModel):
    """認可開始リクエスト。

    Note: authorize_url, token_url, redirect_uri は HttpUrl 型でバリデーションされます。
    不正なURL形式の場合は422エラーを返します。HTTP/HTTPSスキームのみ許可されます。
    """

    server_id: str = Field(..., description="対象サーバーID")
    scopes: List[str] = Field(default_factory=list, description="要求スコープ")
    authorize_url: Optional[HttpUrl] = Field(
        default=None, description="OAuth authorize endpoint URL (override, HTTP/HTTPS required)"
    )
    token_url: Optional[HttpUrl] = Field(
        default=None, description="OAuth token endpoint URL (override, HTTP/HTTPS required)"
    )
    client_id: Optional[str] = Field(
        default=None, description="OAuth client_id (override)"
    )
    redirect_uri: Optional[HttpUrl] = Field(
        default=None, description="OAuth redirect_uri (override, HTTP/HTTPS required)"
    )
    code_challenge: Optional[str] = Field(
        default=None,
        description="クライアント生成の PKCE code_challenge",
    )
    code_challenge_method: str = Field(
        default="S256",
        description="PKCE code_challenge_method (例: S256/plain)",
    )


class OAuthInitiateResponse(BaseModel):
    """認可開始レスポンス。"""

    auth_url: str = Field(..., description="リダイレクト先認可URL")
    state: str = Field(..., description="CSRF 防止用 state")
    required_scopes: List[str] = Field(default_factory=list, description="要求スコープ")


class OAuthCallbackResponse(BaseModel):
    """認可コールバックレスポンス。"""

    status: str = Field(..., description="認可状態")
    scope: List[str] = Field(default_factory=list, description="認可されたスコープ")
    expires_in: Optional[int] = Field(default=None, description="アクセストークン有効秒数")
    credential_key: Optional[str] = Field(
        default=None, description="保存された資格情報の参照キー"
    )
    expires_at: Optional[datetime] = Field(
        default=None, description="アクセストークンの有効期限（UTC）"
    )


class OAuthRefreshRequest(BaseModel):
    """トークンリフレッシュリクエスト。"""

    server_id: str = Field(..., description="対象サーバーID")
    credential_key: str = Field(..., description="リフレッシュ対象の credential_key")


class OAuthRefreshResponse(BaseModel):
    """トークンリフレッシュレスポンス。"""

    credential_key: str = Field(..., description="新しい credential_key")
    refreshed: bool = Field(..., description="リフレッシュが実行されたか")
    scope: List[str] = Field(default_factory=list, description="付与されたスコープ")
    expires_at: datetime = Field(..., description="アクセストークンの有効期限（UTC）")

File: frontend/lib/api/oauth.ts
// OAuth 関連 API クライアント
import { CatalogItem } from '../types/catalog';

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000';

export interface OAuthInitiatePayload {
  serverId: string;
  scopes: string[];
  authorizeUrl?: string;
  tokenUrl?: string;
  clientId?: string;
  redirectUri?: string;
  codeChallenge?: string;
  codeChallengeMethod?: string;
}

export interface OAuthInitiateResult {
  auth_url: string;
  state: string;
  required_scopes: string[];
}

export interface OAuthExchangeParams {
  code: string;
  state: string;
  serverId: string;
  codeVerifier?: string;
}

export interface OAuthExchangeResult {
  status: string;
  scope: string[];
  credential_key?: string;
  expires_at?: string;
}

export interface OAuthRefreshParams {
  serverId: string;
  credentialKey: string;
}

export interface OAuthRefreshResult {
  credential_key: string;
  refreshed: boolean;
  scope: string[];
  expires_at: string;
}

function toQuery(params: Record<string, string | undefined | null>): string {
  const query = new URLSearchParams();
  Object.entries(params).forEach(([key, value]) => {
    if (value !== undefined && value !== null) {
      query.append(key, value);
    }
  });
  return query.toString();
}

export async function initiateOAuth(payload: OAuthInitiatePayload): Promise<OAuthInitiateResult> {
  const response = await fetch(`${API_BASE_URL}/api/catalog/oauth/initiate`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      server_id: payload.serverId,
      scopes: payload.scopes,
      authorize_url: payload.authorizeUrl,
      token_url: payload.tokenUrl,
      client_id: payload.clientId,
      redirect_uri: payload.redirectUri,
      code_challenge: payload.codeChallenge,
      code_challenge_method: payload.codeChallengeMethod ?? 'S256',
    }),
  });

  if (!response.ok) {
    const error = await response.json().catch(() => ({}));
    throw new Error(error.message || '認可の開始に失敗しました');
  }
  return response.json();
}

export async function exchangeOAuth(params: OAuthExchangeParams): Promise<OAuthExchangeResult> {
  const query = toQuery({
    code: params.code,
    state: params.state,
    server_id: params.serverId,
    code_verifier: params.codeVerifier,
  });
  const response = await fetch(`${API_BASE_URL}/api/catalog/oauth/callback?${query}`);
  if (!response.ok) {
    const error = await response.json().catch(() => ({}));
    throw new Error(error.message || 'トークン交換に失敗しました');
  }
  return response.json();
}

export async function refreshOAuth(params: OAuthRefreshParams): Promise<OAuthRefreshResult> {
  const response = await fetch(`${API_BASE_URL}/api/catalog/oauth/refresh`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      server_id: params.serverId,
      credential_key: params.credentialKey,
    }),
  });
  if (!response.ok) {
    const error = await response.json().catch(() => ({}));
    throw new Error(error.message || 'トークンリフレッシュに失敗しました');
  }
  return response.json();
}

export function collectScopes(item: CatalogItem): string[] {
  return item.required_scopes ?? [];
}
